### PR TITLE
[Fix #4444] Fix `Style/AutoResourceCleanup` shouldn't flag `File.open(...).close

### LINKED ISSUE #4444: `Style/AutoResourceCleanup` shouldn't flag `File.open(…).close`.
The `Style/AutoResourceCleanup` cop should not flag the following (`touch` + `chmod`):

```ruby
File.open('file', 'w', 0o777).close
```

--------

## Expected behavior

Given a file with the content above, `rubocop --only Style/AutoResourceCleanup` should output:

```
Inspecting 1 file
.

1 file inspected, no offenses detected
```

## Actual behavior

Given a file with the content above, `rubocop --only Style/AutoResourceCleanup` outputs:

```
Inspecting 1 file
C

Offenses:

open.rb:1:1: C: Use the block version of File.open.
File.open(file, 'w', 0o100777).close
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected
```

## RuboCop version

```
$ rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.0.0 universal.x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by dpostorivo**
This commit addresses #4444 by checking if the function to clean up the resource is immediately cleaned up.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: spec/rubocop/cop/style/auto_resource_cleanup_spec.rb [hunk 1]
```diff
@@ -17,4 +17,8 @@
   it 'does not register an offense for File.open with block-pass' do
     expect_no_offenses('File.open("file", &:read)')
   end
+
+  it 'does not register an offense for File.open with immideate close' do
```
Comment (nsemrau) [inline spec/rubocop/cop/style/auto_resource_cleanup_spec.rb]: s/immideate/immediate :)

**Comment by bbatsov** [comment]
Seems to me that a better way to handle this would be to actually check that the result of `File.open` is assigned to some variable or is the return of some method. And I guess that should start using the node pattern matching approach now.

File: lib/rubocop/cop/style/auto_resource_cleanup.rb [hunk 2]
```diff
@@ -23,21 +23,34 @@ class AutoResourceCleanup < Cop
...
                                         method: target_method))
           end
         end
+
+        private
+
+        def cleanup?(node, target_class)
+          parent = node.parent
+          cleanup_method = CLEANUP_METHODS[target_class]
+          return true if parent && parent.block_type?
```
Comment (rrosenblum) [inline lib/rubocop/cop/style/auto_resource_cleanup.rb]: What are your thoughts on combining this into a single statement?

```ruby
node.block_argument? ||
  (parent && (parent.block_type? || parent.method_name == cleanup_method))
```

Comment (dpostorivo) [reply lib/rubocop/cop/style/auto_resource_cleanup.rb]: :+1:

**Pull request merged**